### DISCUSSION TITLE
Best way to add WebRTC stream to Nicegui app

### DISCUSSION THREAD

**Discussion by h0uter** [zauberzeug/nicegui#3147] [Ideas / Feature Requests]
What is the recommended way for displaying a WebRTC stream in Nicegui?

I found this Nicegui example that can display a webcam (and rtsp?) stream using OpenCV: https://github.com/zauberzeug/nicegui/blob/main/examples/opencv_webcam/main.py. However that approach does not appear to support WebRTC.

Upon further search I found this related solution for Streamlit: https://github.com/whitphx/streamlit-webrtc

Another potential approach could be embedding this javascript example from the aiortc WebRTC Python package in a `ui.html` element: https://github.com/aiortc/aiortc/tree/main/examples/server

I would be interested to hear ideas on the best approach to add a WebRTC stream to a nicegui app! This would be really usefull for robotics applications.

**Comment by Dronakurl** [comment]
Did you achieve this somehow? I desparately need this feature

**Comment by rodja** [comment]
What kind of features do you need from WebRTC? If you only nee the video transfer, you can look at [`ui.interactive_image`](https://nicegui.io/documentation/interactive_image#interactive_image). It automatically adapts frame updates to the available bandwidth per client.

**Reply by Dronakurl** [reply]
I would like to have an element on my site which triggers a "allow to use my webcam" request. Then the webcam captures the video stream and sends it back to the server. `interactive_image` is the other way round.

**Reply by rodja** [reply]
I totally understand that WebRTC would be cool. But have you seen our opencv webcam example: https://github.com/zauberzeug/nicegui/blob/main/examples/opencv_webcam/main.py
If you add a button to show/hide the `ui.interactive_image` it should be fairly similar to what you describe...

**Reply by Dronakurl** [reply]
Thank's for the link. I'm already using opencv and that example helped a lot to get started, but I don't wan't to use a webcam that is connected to the server, but a webcam provided by a browser client connecting to my nicegui app. That's the promise of streamlit-webrtc (which I haven't checked out yet).

**Reply by rodja** [reply]
I think you can also set a network source in OpenCV. That way the NiceGUI server will just be a relay. Don't know if streamlit-webrtc supports such a scenario (eg. when the browser is not in the same network as the camera).

**Reply by Dronakurl** [reply]
You can set a network source in opencv, e.g. rtsp cameras, but how to get a network stream from a camera provided by a browser client?

**Reply by rodja** [reply]
Hmm.. so your camera does not provide a rtsp stream but only WebRTC? Them maybe you can utilize https://github.com/aiortc/aiortc or similar to get the frames in Python.

**Reply by rodja** [reply]
Or do you want to share a camera from the client (eg. the browser visiting the server) to send the data to the server? Like in a video call?

**Reply by Dronakurl** [reply]
Yes, exactly like in a video call. A user would call up the nicegui application in his browser and will be asked to share his camera. The camera frames are sent to the server and then transformed/ analyzed.

**Reply by rodja** [reply]
Ok, there is already a feature request for capturing the user's camera image: https://github.com/zauberzeug/nicegui/discussions/2010. It got a bit stale but you could try out whats suggested there and pick up the discussion.

**Reply by Dronakurl** [reply]
Thank you, that works for me!

**Comment by frankvp11** [comment]
Has there been any updates on this?
I'm quite interested in using WebRTC / aiortc to support a calling like feature in NiceGUI. 
If there are any code snippets I could look to, that would be great. 
Edit:
I've found this example for FastAPI and i'm quite curious to know if it would work for NiceGUI. 
https://github.com/priyanshu88064/video-calling-app
Granted, I have not yet tested it (busy with exams :cry:)
